apiserver/facades/client/storage/storage.go

```diff
@@ -891,3 +892,124 @@ func (a *APIv3) Attach(args params.StorageAttachmentIds) (params.ErrorResults, e
+			continue
+		}
+		results[i].Result = details
+	}
+	return params.ImportStorageResults{Results: results}, nil
+}
+
+func (a *APIv4) importStorage(arg params.ImportStorageParams) (*params.ImportStorageDetails, error) {
+	if arg.Kind != params.StorageKindFilesystem {
+		// TODO(axw) implement support for volumes.
+		return nil, errors.NotSupportedf("storage kind %q", arg.Kind.String())
+	}
+	if !storage.IsValidPoolName(arg.Pool) {
+		return nil, errors.NotValidf("pool name %q", arg.Pool)
+	}
+
+	cfg, err := a.poolManager.Get(arg.Pool)
+	if errors.IsNotFound(err) {
+		cfg, err = storage.NewConfig(
+			arg.Pool,
+			storage.ProviderType(arg.Pool),
+			map[string]interface{}{},
+		)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	} else if err != nil {
+		return nil, errors.Trace(err)
+	}
+	provider, err := a.registry.StorageProvider(cfg.Provider())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return a.importFilesystem(arg, provider, cfg)
+}
+
+func (a *APIv4) importFilesystem(
+	arg params.ImportStorageParams,
+	provider storage.Provider,
+	cfg *storage.Config,
+) (*params.ImportStorageDetails, error) {
+	resourceTags := map[string]string{
+		tags.JujuModel:      a.storage.ModelTag().Id(),
+		tags.JujuController: a.storage.ControllerTag().Id(),
+	}
+	var volumeInfo *state.VolumeInfo
+	filesystemInfo := state.FilesystemInfo{Pool: arg.Pool}
+
+	// If the storage provider supports filesystems, import the filesystem,
+	// otherwise import a volume which will back a filesystem.
+	if provider.Supports(storage.StorageKindFilesystem) {
+		filesystemSource, err := provider.FilesystemSource(cfg)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		filesystemImporter, ok := filesystemSource.(storage.FilesystemImporter)
+		if !ok {
+			return nil, errors.NotSupportedf(
+				"importing filesystem with storage provider %q",
+				cfg.Provider(),
+			)
+		}
+		info, err := filesystemImporter.ImportFilesystem(arg.ProviderId, resourceTags)
+		if err != nil {
+			return nil, errors.Annotate(err, "importing filesystem")
+		}
+		filesystemInfo.FilesystemId = arg.ProviderId
+		filesystemInfo.Size = info.Size
+	} else {
+		volumeSource, err := provider.VolumeSource(cfg)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		volumeImporter, ok := volumeSource.(storage.VolumeImporter)
+		if !ok {
+			return nil, errors.NotSupportedf(
+				"importing volume with storage provider %q",
+				cfg.Provider(),
+			)
+		}
+		info, err := volumeImporter.ImportVolume(arg.ProviderId, resourceTags)
+		if err != nil {
+			return nil, errors.Annotate(err, "importing volume")
+		}
+		volumeInfo = &state.VolumeInfo{
+			HardwareId: info.HardwareId,
+			WWN:        info.WWN,
+			Size:       info.Size,
+			Pool:       arg.Pool,
+			VolumeId:   info.VolumeId,
+			Persistent: info.Persistent,
+		}
+		filesystemInfo.Size = info.Size
+	}
+
+	storageTag, err := a.storage.ImportFilesystem(filesystemInfo, volumeInfo, arg.StorageName)
```

Comment:
There are a confusing number of things called `ImportFilesystem` in this method (and in this change I guess). I feel like it would be clearer if we distinguish the names a bit - all of the steps are part of importing a filesystem, but it would be easier to grab onto pieces of it if they had different names. 

The state method is really adding a filesystem (with some details indicating it's preexisting) - could you change it to be `AddFilesystem` or `AddExistingFilesystem`?